string/src/io/github/iltotore/iron/string/constraint.scala

```diff
@@ -12,40 +13,53 @@ object constraint {
    */
   trait LowerCase
 
-  inline given Constraint.RuntimeOnly[String, LowerCase] with {
-    override inline def assert(value: String): Boolean = value equals value.toLowerCase
+  inline given Constraint[String, LowerCase] with {
+    override inline def assert(value: String): Boolean = compileTime.checkLowerCase(value)
 
-    override inline def getMessage(value: String): String = s"$value should be lower cased"
+    override inline def getMessage(value: String): String = value + " should be lower cased"
   }
 
   /**
    * Constraint: checks if the input value is upper case.
    */
   trait UpperCase
 
-  inline given Constraint.RuntimeOnly[String, UpperCase] with {
-    override inline def assert(value: String): Boolean = value equals value.toUpperCase
+  inline given Constraint[String, UpperCase] with {
+    override inline def assert(value: String): Boolean = compileTime.checkUpperCase(value)
 
-    override inline def getMessage(value: String): String = s"$value should be upper cased"
+    override inline def getMessage(value: String): String = value + " should be upper cased"
   }
 
   /**
    * Constraint: checks if the input value matches V.
    * @tparam V the regex to match with.
    */
-  trait Match[V]
+  trait Match[V <: String & Singleton]
 
   type Alphanumeric = Match["^[a-zA-Z0-9]+"] DescribedAs "Value should be alphanumeric"
 
+  inline given MatchConstraint[Alphanumeric] with {}
+
   type URLLike =
     Match["^(?:http(s)?:\\/\\/)?[\\w.-]+(?:\\.[\\w\\.-]+)+[\\w\\-\\._~:/?#\\[\\]@!\\$&'\\(\\)\\*\\+,;=.]+$"] DescribedAs "Value should be an URL"
 
+  inline given MatchConstraint[URLLike] with {}
+
   type UUIDLike =
     Match["^([0-9a-fA-F]{8}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{12})"] DescribedAs "Value should be an UUID"
 
-  inline given [V <: String]: Constraint.RuntimeOnly[String, Match[V]] with {
+  inline given MatchConstraint[UUIDLike] with {}
+
+  trait MatchConstraint[M <: Match[_] | DescribedAs[Match[_], _]] extends Constraint[String, M] {
```

Comment:
You shouldn't have to precise the `DescribedAs[Match[_], _]]` since ad-hoc polymorphism will do this for you. See the `DescribedAs` definition in `io.github.iltotore.iron.constraint`

Reply:
It unfortunatelly does not work: 
```Type argument io.github.iltotore.iron.string.constraint.Alphanumeric does not conform to upper bound io.github.iltotore.iron.string.constraint.Match[?]```

I guess that because typeclasses work on objects of specific types, but not types themselves.

Reply:
I think this is due to your other inline given. What if you structure your Match constraint like this ?
```scala
class MatchConstraint[V <: String] extends Constraint[String, Match[V]] {

  override inline def assert(value: String): Boolean = compileTime.checkMatch(value, constValue[V])

  override inline def getMessage(value: String): String = value + " should match " + constValue[V]
} 

inline given [V <: String]: MatchConstraint[V] = new MatchConstraint
```

This simply should work, assuming your `checkMatch` method handles runtime (uncomputable at compile-time) values.

Reply:
I'm not sure it will work - I don't think `new` will be invoked in compile time to be honest.

Reply:
That's how other constraints are built so it should work.

Reply:
So @andrzejressel I just tested and the sample I posted above works as expected. Here is how the `Match` implementation looks:
```scala
  /**
   * Constraint: checks if the input value matches V.
   * @tparam V the regex to match with.
   */
  trait Match[V]

  type Alphanumeric = Match["^[a-zA-Z0-9]+"] DescribedAs "Value should be alphanumeric"

  type URLLike =
    Match["^(?:http(s)?:\\/\\/)?[\\w.-]+(?:\\.[\\w\\.-]+)+[\\w\\-\\._~:/?#\\[\\]@!\\$&'\\(\\)\\*\\+,;=.]+$"] DescribedAs "Value should be an URL"

  type UUIDLike =
    Match["^([0-9a-fA-F]{8}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{12})"] DescribedAs "Value should be an UUID"

  class MatchConstraint[V <: String] extends Constraint[String, Match[V]] {

    override inline def assert(value: String): Boolean = compileTime.checkMatch(value, constValue[V])

    override inline def getMessage(value: String): String = "regex"
  }

  inline given [V <: String]: MatchConstraint[V] = new MatchConstraint
```

I suggest you to apply this to your PR (and do the same for LowerCase and UpperCase).

Reply:
That's so weird ...

So inline given with does not work with type argument, but inline given with new does?

I'll try to investigate and report it to dotty, because I don't think it should work that way

Reply:
I've fixed Match implementation, but I don't understand what should I do with LowerCase and UpperCase - they work fine already